api/src/main/java/org/openmrs/api/impl/TagServiceImpl.java

```diff
@@ -0,0 +1,183 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.api.impl;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.openmrs.OpenmrsObject;
+import org.openmrs.api.APIException;
+import org.openmrs.Tag;
+import org.openmrs.api.TagService;
+import org.openmrs.api.context.Context;
+import org.openmrs.api.db.TagDAO;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.ArrayList;
+
+@Transactional(readOnly = true)
+public class TagServiceImpl extends BaseOpenmrsService implements TagService {
+	
+	protected final Log log = LogFactory.getLog(this.getClass());
+	
+	private TagDAO dao;
+	
+	/**
+	 * Injected in moduleApplicationContext.xml
+	 */
+	public void setDao(TagDAO dao) {
+		this.dao = dao;
+	}
+	
+	@Override
+	public Tag getTagByUuid(String uuid) {
+		return dao.getTagByUuid(uuid);
+	}
+	
+	@Override
+	@Transactional(readOnly = false)
+	public Tag saveTag(Tag tag) {
+		return dao.saveTag(tag);
+	}
+	
+	@Override
+	@Transactional(readOnly = false)
+	public void deleteTag(Tag tag) {
+		dao.deleteTag(tag);
+	}
+	
+	@Override
+	@Transactional(readOnly = false)
+	public void removeTag(OpenmrsObject openmrsObject, String tag) {
+		Tag tag1 = dao.getTag(openmrsObject, tag);
+		if (tag1.equals(null)) {
+			log.warn("Tag does not exist");
+		} else {
+			deleteTag(tag1);
+		}
+	}
+	
+	@Override
+	public List<String> getAllTags() {
+		return dao.getAllTags();
+	}
+	
+	@Override
+	public Tag getTag(Integer tagId) {
+		return dao.getTag(tagId);
+	}
+	
+	@Override
+	public List<Tag> getTags(String tag) {
+		return dao.getTags(tag);
+	}
+	
+	@Override
+	public <T extends OpenmrsObject> T getObject(Class<T> objectType, String objectUuid) {
+		Object object;
+		object = dao.getObject(objectUuid, objectType.getName());
+		return objectType.cast(object);
```

Comment:
Why do you have to cast? You need to change getObject in the DAO to take in a class object too, that way you don't have to cast here, and you will need to cast using (T) and not calling cast